src/rttable.c

```diff
@@ -41,6 +41,7 @@
 #include "igmpproxy.h"
 
 #define MAX_ORIGINS 4
+#define MAX_DOWNSTREAM_HOSTS 20
```

Comment:
No the last host tracking still works with more than 20 downstream hosts.

Agree the exact number of 20 is a bit arbitary however I will explain the idea behind limiting it. The goal is just to avoid leaving a stream while still downstream hosts are joined, in order to do that in fact I only have to know if at least one other downstream host is watching. 

If I have 20 registered downstream hosts I'm pretty sure at least more than 1 is watching, certainly because when the 20 slot's are filled I start overwriting them with new joins that come in. The chance that all registered hosts in the route have left while an overwritten downstream hosts is still watching is with 20 registered hosts almost nil.

The alternative is to register all downstream hosts indeed but in very large deployments with a lot of downstream hosts that in my opinion is a waste of resources because there is no need to know all of them, it does not add anything from my point of view.

Reply:
Thank you for explanation, I understood it. igmpproxy just needs to know if there is at least one downstream host for registered group, exact number is not important and also list of downstream host is not needed for this functionality.

But for larger deployments even this your proposed solution may cause problems. There is a space for 20 hosts and you store host into first free slot. If there is not free slot then you always overwrite host stored in slot zero. So if 40 (or mote hosts) register for one multicast group then first 20 hosts are stored into table and later all other hosts are sequentially overwriting slot zero, with result that last host is stored in slot zero. Now when hosts 40 (last one), 2, 3, 4, ..., 20 leave the group then igmpproxy would think that there is no downstream host registered. For larger deployments there can be even more downstream hosts (lets say 200 or more) and it can be possible that "problematic set" would leave group in same time. Probability increases with larger and larger deployments.

Moreover this algorithm is fully deterministic, so "bad clients" can be specially created to trigger this problem. For example if you have 40 downstream hosts and 20 would cooperate, then they can break whole multicast streaming for other hosts. This even does not need large deployments, just 20 hosts are needed.

Reply:
I was thinking a bit more about this problem. And basically `quickleave` mode is incompatible with IGMP protocol on a network with more then one downstream host.

According to IGMPv2 specification, RFC 2236, downstream host does not have to send join or leave message when there is another downstream host on network which is connected to same multicast group. So whole "reference counting" is broken/incompatible with IGMPv2.

But I understand needs for `quickleave`, specially for high bitrate IPTV streams on slow link from ISP, e.g. when link is able to receive N streams in parallel, but not N+1 (e.g. for slow DSL N can be 2 or 3). And in this case IPTV boxes, specially designed by ISP, would not do above "optimization" (send join/leave message only if there is not somebody else already joined), but rather send join/leave messages always and immediately to ensure that link is not exhausted.

So in this specific case we can expect that join/leave messages are always send by clients (or at least join messages to ensure that we did not disconnect any client). But whole downstream igmp/multicast network would be broken when there come some clients which do not do this.

Heh, by broken I mean that some mutlicast groups can be left on upstream even when there are downstream clients; but as igmpproxy is doing periodic time-based query of downstream hosts, so when timeout expires igmpproxy would know that there are downstream clients and join to those multicast groups on upstream again.

I would think more about solution for `quickleave` mode to fix problem which you are trying to fix with in mind that it is incompatible with IGMPv2 specification and therefore clients can broke it...

If you come up with some other ideas, please let me know.

Reply:
From my professional experience with telecom IGMP devices performing "proxy mode" (not snooping mode) always do exactly the thing I implemented with this pull request. So they track all downstream hosts and only send Join or Leave messages upstream when the first or the last hosts joins or leaves a stream. 

But if I understand you right your problem with it is that it's not directly related to quickleave mode and that therefore it feels wrong to completely change the behavior of quickleave mode? 

Maybe it's an option to make a seperate configuration option for this in the config file? Something like quickleave-hostbased? So that will mean with only quickleave in the config file we keep the current behavior and with quicleave enabled and also the quickleave-hostbased option set we use the host tracking I propose.

Because there definitely is some relation with quickleave mode, when quickleave mode is disabled you won't have this problem anyway. So host tracking without quickleave mode makes no sense.

Reply:
> From my professional experience with telecom IGMP devices performing "proxy mode" (not snooping mode) always do exactly the thing I implemented with this pull request. So they track all downstream hosts and only send Join or Leave messages upstream when the first or the last hosts joins or leaves a stream.

You do this just for *first 20* hosts, not for tracking *all* hosts.

> But if I understand you right your problem with it is that it's not directly related to quickleave mode and that therefore it feels wrong to completely change the behavior of quickleave mode?

No, this is directly related to quickleave mode.

I described in previous post a problem that it is not compatible with IGMP protocol specification even if you even track *all* hosts in your implementation. (But your implementation does not track all hosts, but only 20).

But also I wrote that in that specific scenario it is the best what can be done for quickleave mode to track join/leave messages in downstream hosts. In case you do not have better idea how to do it. If you wrote that telecom IGMP devices are doing same thing then we probably do not come up with something better.

So the main remaining part is what to do with limitation that this proposed change tracks only (some random) 20 hosts and not all hosts.